source/common/common/fancy_logger.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include <string>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/synchronization/mutex.h"
+#include "spdlog/spdlog.h"
+
+namespace Envoy {
+
+using FancyMap = absl::flat_hash_map<std::string, std::shared_ptr<spdlog::logger>>;
+using FancyMapPtr = std::shared_ptr<FancyMap>;
+using SpdLoggerPtr = std::shared_ptr<spdlog::logger>;
+
+/**
+ * Stores the lock and functions used by Fancy Logger's macro so that we don't need to declare
+ * them globally. Functions are provided to initialize a logger, set log level, flush a logger.
+ */
+class FancyContext {
+public:
+  /**
+   * Gets a logger from map given the key (e.g. file name).
+   */
+  static SpdLoggerPtr getFancyLogEntry(std::string key) ABSL_LOCKS_EXCLUDED(fancy_log_lock_);
+
+  /**
+   * Initializes Fancy Logger and register it in global map if not done.
+   */
+  static void initFancyLogger(std::string key, std::atomic<spdlog::logger*>& logger)
+      ABSL_LOCKS_EXCLUDED(fancy_log_lock_);
+
+  /**
+   * Sets log level. If not found, return false.
+   */
+  static bool setFancyLogger(std::string key, spdlog::level::level_enum log_level)
+      ABSL_LOCKS_EXCLUDED(fancy_log_lock_);
+
+  /**
+   * Sets the default logger level and format when updating context.
+   */
+  static void setDefaultFancyLevelFormat(spdlog::level::level_enum level, std::string format)
+      ABSL_LOCKS_EXCLUDED(fancy_log_lock_);
+
+private:
+  /**
+   * Initializes sink for the initialization of loggers, needed only in benchmark test.
+   */
+  static void initSink();
+
+  /**
+   * Creates a logger given key and log level, and add it to map.
+   * Key is the log component name, e.g. file name now.
+   */
+  static spdlog::logger* createLogger(std::string key, int level = -1)
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(fancy_log_lock_);
+
+  /**
+   * Lock for the following map (not for the corresponding loggers).
+   */
+  ABSL_CONST_INIT static absl::Mutex fancy_log_lock_;
+
+  /**
+   * Map that stores <key, logger> pairs, key can be the file name.
+   */
+  static FancyMapPtr fancy_log_map_ ABSL_GUARDED_BY(fancy_log_lock_);
```

Comment:
Those are likely cause static initializations fiasco, can you make those non static and make it singleton? (use MUTABLE_CONSTRUCT_ON_FIRST_USE)

Reply:
Addressed now.